sse2neon.h

```diff
@@ -8358,6 +8359,530 @@ FORCE_INLINE int _mm_testz_si128(__m128i a, __m128i b)
+    }
+    unsigned short *ptr = (unsigned short *) mtx;
+    for (int i = 0; i < bound; i++) {
+        int val = 1;
+        for (int j = 0, k = i; j < bound - i && k < bound; j++, k++)
+            val &= ptr[k * bound + j];
+        res = (val << i) + res;
+    }
+    return res;
+}
+
+static int _sse2neon_cmp_byte_equal_ordered(__m128i a,
+                                            int la,
+                                            __m128i b,
+                                            int lb)
+{
+    __m128i mtx[16];
+    PCMPSTR_EQ_8x16(a, b, mtx);
+    return aggregate_equal_ordered_8x16(16, la, lb, mtx);
+}
+
+static int _sse2neon_cmp_word_equal_ordered(__m128i a,
+                                            int la,
+                                            __m128i b,
+                                            int lb)
+{
+    __m128i mtx[16];
+    PCMPSTR_EQ_16x8(a, b, mtx);
+    return aggregate_equal_ordered_16x8(8, la, lb, mtx);
+}
+
+typedef int (*cmpestr_func_t)(__m128i a, int la, __m128i b, int lb);
+
+static cmpestr_func_t _sse2neon_cmpfunc_table[] = {
+    /* CMP_UBYTE_EQUAL_ANY */ _sse2neon_cmp_byte_equal_any,
+    /* CMP_UWORD_EQUAL_ANY */ _sse2neon_cmp_word_equal_any,
+    /* CMP_SBYTE_EQUAL_ANY */ _sse2neon_cmp_byte_equal_any,
+    /* CMP_SWORD_EQUAL_ANY */ _sse2neon_cmp_word_equal_any,
+    /* CMP_UBYTE_RANGES */ _sse2neon_cmp_ubyte_ranges,
+    /* CMP_UWORD_RANGES */ _sse2neon_cmp_uword_ranges,
+    /* CMP_SBYTE_RANGES */ _sse2neon_cmp_sbyte_ranges,
+    /* CMP_SWORD_RANGES */ _sse2neon_cmp_sword_ranges,
+    /* CMP_UBYTE_EQUAL_EACH */ _sse2neon_cmp_byte_equal_each,
+    /* CMP_UWORD_EQUAL_EACH */ _sse2neon_cmp_word_equal_each,
+    /* CMP_SBYTE_EQUAL_EACH */ _sse2neon_cmp_byte_equal_each,
+    /* CMP_SWORD_EQUAL_EACH] */ _sse2neon_cmp_word_equal_each,
+    /* CMP_UBYTE_EQUAL_ORDERED */ _sse2neon_cmp_byte_equal_ordered,
+    /* CMP_UWORD_EQUAL_ORDERED */ _sse2neon_cmp_word_equal_ordered,
+    /* CMP_SBYTE_EQUAL_ORDERED */ _sse2neon_cmp_byte_equal_ordered,
+    /* CMP_SWORD_EQUAL_ORDERED */ _sse2neon_cmp_word_equal_ordered};
+
+static inline int neg_fun(int res, int lb, int imm8, int bound)
+{
+    switch (imm8 & 0x30) {
+    case _SIDD_NEGATIVE_POLARITY:
+        res ^= 0xffffffff;
+        break;
+    case _SIDD_MASKED_NEGATIVE_POLARITY:
+        res ^= (1 << lb) - 1;
+        break;
+    default:
+        break;
+    }
+
+    return res & ((bound == 8) ? 0xFF : 0xFFFF);
+}
+
+#define SSE2NEON_MAX(x, y) (x) > (y) ? (x) : (y)
+
+// Compare packed strings in a and b with lengths la and lb using the control
+// in imm8, and store the generated index in dst.
+// https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestri
+FORCE_INLINE int _mm_cmpestri(__m128i a,
+                              int la,
+                              __m128i b,
+                              int lb,
+                              const int imm8)
+{
+    int upper = (imm8 & 0x01) ? 8 : 16;
+#if defined(__aarch64__)
```

Comment:
Can you check if generic C code benefits from compiler optimizations (at least `-O2`) which generate instructions better than hand-written inline assembly?

Reply:
Got it. I will post the assembly in this conversation.

Reply:
Good. The original inline assembly listing can be replaced with generic C implementation once we confirm the effectiveness of compiler optimizations.

Reply:
After the experiment on Compiler Explorer, in brief, generic C implementation outputs less assembly codes compared to hand-written inline assembly with `-O2` compiler optimization flag (there are no any differences when applying `-O3` flag, so I will  only post the generated assembly with `-O2` flag).

The test program is written as follows:

<table>
<tr>
<td> hand-written inline assembly </td> <td> generic C implementation </td>
</tr>
<tr>
<td>

```c
#include <stdio.h>

#define SSE2NEON_MAX(x, y) (x) > (y) ? (x) : (y)

int main() {
    int la = 5, lb = 10, upper = 16;
    __asm__ __volatile__(
        "eor w0, %w[a], %w[a], asr #31\n"
        "sub %w[a], w0, %w[a], asr #31\n"
        "eor w1, %w[b], %w[b], asr #31\n"
        "sub %w[b], w1, %w[b], asr #31\n"
        "cmp %w[a], %w[bd]\n"
        "csel %w[a], %w[bd], %w[a], gt\n"
        "cmp %w[b], %w[bd]\n"
        "csel %w[b], %w[bd], %w[b], gt\n"
        : [a] "+r"(la), [b] "+r"(lb)
        : [bd] "r"(upper)
        : "w0", "w1");

    printf("%d %d", la, lb);

    return 0;
}
```

</td>
<td>

```c
#include <stdio.h>

#define SSE2NEON_MAX(x, y) (x) > (y) ? (x) : (y)

int main() {
    int la = 5, lb = 10, upper = 16;
    int tmp1 = la ^ (la >> 31);
    la = tmp1 - (la >> 31);
    int tmp2 = lb ^ (lb >> 31);
    lb = tmp2 - (lb >> 31);
    la = SSE2NEON_MAX(la, upper);
    lb = SSE2NEON_MAX(lb, upper);

    printf("%d %d", la, lb);

    return 0;
}
```

</td>
</tr>
</table>

And the results of GCC 10.2 (Compiler Explorer does not provide GCC 10, so I use GCC 10.2 instead):
<table>
<tr>
<td> hand-written inline assembly </td> <td> generic C implementation </td>
</tr>
<tr>
<td>

```assembly
.LC0:
        .string "%d %d"
main:
        stp     x29, x30, [sp, -16]!
        mov     w3, 5
        mov     w2, 10
        mov     x29, sp
        mov     w4, 16
        eor w0, w3, w3, asr #31
sub w3, w0, w3, asr #31
eor w1, w2, w2, asr #31
sub w2, w1, w2, asr #31
cmp w3, w4
csel w3, w4, w3, gt
cmp w2, w4
csel w2, w4, w2, gt

        adrp    x0, .LC0
        mov     w1, w3
        add     x0, x0, :lo12:.LC0
        bl      printf
        mov     w0, 0
        ldp     x29, x30, [sp], 16
        ret
```

</td>
<td>

```assembly
.LC0:
        .string "%d %d"
main:
        stp     x29, x30, [sp, -16]!
        mov     w2, 16
        adrp    x0, .LC0
        mov     x29, sp
        mov     w1, w2
        add     x0, x0, :lo12:.LC0
        bl      printf
        mov     w0, 0
        ldp     x29, x30, [sp], 16
        ret
```

</td>
</tr>
</table>

Also the results of Clang 11.0.0:
<table>
<tr>
<td> hand-written inline assembly </td> <td> generic C implementation </td>
</tr>
<tr>
<td>

```assembly
main:                                   // @main
        stp     x29, x30, [sp, #-16]!           // 16-byte Folded Spill
        mov     x29, sp
        mov     w8, #5
        mov     w2, #10
        mov     w9, #16
        eor     w0, w8, w8, asr #31
        sub     w8, w0, w8, asr #31
        eor     w1, w2, w2, asr #31
        sub     w2, w1, w2, asr #31
        cmp     w8, w9
        csel    w8, w9, w8, gt
        cmp     w2, w9
        csel    w2, w9, w2, gt

        adrp    x0, .L.str
        add     x0, x0, :lo12:.L.str
        mov     w1, w8
        bl      printf
        mov     w0, wzr
        ldp     x29, x30, [sp], #16             // 16-byte Folded Reload
        ret
.L.str:
        .asciz  "%d %d"
```

</td>
<td>

```assembly
main:                                   // @main
        stp     x29, x30, [sp, #-16]!           // 16-byte Folded Spill
        mov     x29, sp
        adrp    x0, .L.str
        add     x0, x0, :lo12:.L.str
        mov     w1, #16
        mov     w2, #16
        bl      printf
        mov     w0, wzr
        ldp     x29, x30, [sp], #16             // 16-byte Folded Reload
        ret
.L.str:
        .asciz  "%d %d"
```

</td>
</tr>
</table>

Reply:
It looks promising. Let's drop the use of inline assembly, which does not outperform the generic C code.

Reply:
After testing, I found the generic C implementation will fail on `EQUAL_ORDERED` test cases. I will investigate the root cause.

Reply:
So it seems I misinterpret these lines to `MAX`:
```assembly
"cmp %w[a], %w[bd]\n"
"csel %w[a], %w[bd], %w[a], gt\n"
```

It should be `MIN`, though.

Reply:
After testing, the generic C implementation still gives promising results with `-O2` flag.

Testing program:

```c
#include <stdio.h>

#define SSE2NEON_MIN(x, y) (x) < (y) ? (x) : (y)

int main() {
    int la = 5, lb = 10, upper = 16;
    int tmp1 = la ^ (la >> 31);
    la = tmp1 - (la >> 31);
    int tmp2 = lb ^ (lb >> 31);
    lb = tmp2 - (lb >> 31);
    la = SSE2NEON_MIN(la, upper);
    lb = SSE2NEON_MIN(lb, upper);

    printf("%d %d", la, lb);

    return 0;
}
```

And the assembly generated by each compiler:

<table>
<tr>
<td> ARM64 GCC 10.2 </td> <td> armv8-a clang 11.0.0 </td>
</tr>

<tr>
<td>

```asm
.LC0:
        .string "%d %d"
main:
        stp     x29, x30, [sp, -16]!
        mov     w2, 10
        mov     w1, 5
        mov     x29, sp
        adrp    x0, .LC0
        add     x0, x0, :lo12:.LC0
        bl      printf
        mov     w0, 0
        ldp     x29, x30, [sp], 16
        ret
```

</td>
<td>

```asm
main:                                   // @main
        stp     x29, x30, [sp, #-16]!           // 16-byte Folded Spill
        mov     x29, sp
        adrp    x0, .L.str
        add     x0, x0, :lo12:.L.str
        mov     w1, #5
        mov     w2, #10
        bl      printf
        mov     w0, wzr
        ldp     x29, x30, [sp], #16             // 16-byte Folded Reload
        ret
.L.str:
        .asciz  "%d %d"
```

</td>
</tr>
</table>